pipelines/tekton/azureml-container-pipeline/azureml-container-pipeline.yaml

```diff
@@ -0,0 +1,89 @@
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: azureml-container
+spec:
+  params:
+  - name: model-name
+    type: string
+  - default: "1"
+    name: model-version
+    type: string
+  - name: aws-bucket-name
+    type: string
+  tasks:
+  - name: kserve-download-model
+    params:
+    - name: model-name
+      value: $(params.model-name)
+    - name: current-namespace
+      value: $(context.pipelineRun.namespace)
+    - name: aws-bucket-name
+      value: $(params.aws-bucket-name)
+    taskRef:
+      kind: Task
+      name: kserve-download-model
+    workspaces:
+    - name: workspace
+      workspace: docker
+    - name: aws-secret
+      workspace: aws-secret
+  - name: check-model-exists
+    params:
+    - name: model-name
+      value: $(params.model-name)
+    - name: current-namespace
+      value: $(context.pipelineRun.namespace)
+    runAfter:
+     - kserve-download-model
+    taskRef:
+      kind: Task
+      name: check-model-exists
+    workspaces:
+    - name: workspace
+      workspace: docker
+  - name: create-imagestream
+    params:
+    - name: SCRIPT
+      value: oc create imagestream $(params.model-name)  --lookup-local=true --dry-run=client
+        -o yaml | oc apply -f -
+    - name: VERSION
+      value: latest
+    runAfter:
+     - check-model-exists
+    taskRef:
+      kind: ClusterTask
+      name: openshift-client
+  - name: build-mlflow-container
+    params:
+    - name: IMAGE
+      value: image-registry.openshift-image-registry.svc:5000/$(context.pipelineRun.namespace)/$(params.model-name):$(params.model-version)
+    - name: BUILDER_IMAGE
+      value: registry.redhat.io/rhel8/buildah@sha256:0a86ecbdfbe86e9d225b7fe4b090a0dd6d323f8afdfdf2bd933ff223ddb53320
+    - name: STORAGE_DRIVER
+      value: vfs
+    - name: DOCKERFILE
+      value: ./$(params.model-name)/Containerfile
```

Comment:
One thing that hasn't yet clicked for me after going through this is: how does the Containerfile get from this repository on GitHub, onto this location on disk (PVC?) in the execution context of the buildah clustertask run, so that it can be referenced like this? I don't see it getting downloaded anywhere (I'm probably missing it). Is it part of the model download from S3 or something?

Reply:
Apologies if the naming is confusing, the download is done in [pipelines/tekton/azureml-container-pipeline/kserve-download-model.yaml](https://github.com/piotrpdev/ai-edge/blob/40bcffd163b305c8649f4026e971906cf1b3e2e2/pipelines/tekton/azureml-container-pipeline/kserve-download-model.yaml). This downloads the model files from an S3 storage bucket, which include Containerfiles like the models in the repo do.

[Edgar's Gist](https://gist.github.com/israel-hdez/1848d5a9788140c6f2e42ce316c70d5b) has some more examples of how the download is done, including downloading from Azure (I haven't implemented this yet). If Git is involved, a `git clone` could be made [inside the Containerfile](https://github.com/piotrpdev/azureml-model-to-edge/blob/3356dbf00d80021e6d85f6fd26da9063690b1af9/containerfiles/Containerfile.seldonio.mlserver.mlflow) or using the [`git clone` Tekton Task which is included with OpenShift Pipelines](https://hub.tekton.dev/tekton/task/git-clone)

Reply:
Ah right, I understand now, thanks.

While it will make sense for the model binary file to come from an object store like S3 in many cases, I think we'd typically expect something like a Containerfile to come from a git repo, since it's a source file.

For the future of the demo, if we decide that it's more convenient to download the model from a Git repo, then we'd be getting the Containerfile from there too, so let's leave it as-is for now and talk about it as a future improvement :+1: